### ISSUE TITLE
Bug: `sendUpdate()` not updating parent actor with `predictableActionArguments`

### ISSUE THREAD

**Issue by wlee221** [statelyai/xstate#3533]
### Description

Hello XState team 👋 

My team was working on updating our machine to use `predictableActionArguments`, but found `sendUpdate` to no longer work. This happened when child actor tried to update parent after promise invocation with `onError`. We could confirm this only happens when we set `predictableActionArguments` to `true`.

Please see the [minimal repro](https://codesandbox.io/s/xstate-4-33-2-sendupdate-error-cpjxod?file=/src/actors/child.js) for reference.

### Expected result

`sendUpdate()` should update the parent, and trigger a re-render on `useMachine(...)`.

### Actual result

No re-render happens after `childMachine` makes `onError` transition with `sendUpdate()` action.

### Reproduction

https://codesandbox.io/s/xstate-4-33-2-sendupdate-error-cpjxod?file=/src/actors/child.js

Child actor goes through `idle` -> `submit` -> `idle`, but the parent machine gets updated only for the first two states.

### Additional context

I could confirm that adding `{sync: true}` resolves the error, but we want to avoid this for optimization purposes.

**Comment by Andarist** [comment]
I've investigated this and I see why it happens:

1. child sends an event to the parent, this happens (with the flag) **while** it computes its next state
2. parent computes its next state
3. it's changed so it triggers a rerender through `useSyncExternalStore`'s subscription
4. it's a synchronous rerender so we read the `child.getSnapshot()` that returns an "outdated" value because we didn't yet return back to the child with the control flow and it didn't have a chance to update its internal state

So while I see the issue here and gonna think about potential solutions, I also think that this code should be written differently. Take a look at this simplified demo that works as expected:
https://codesandbox.io/s/xstate-4-33-2-sendupdate-error-forked-zhzm1k?file=/src/index.js:271-333

Basically, you don't need to grab the child's snapshot through the parent - this requires more manual work and requires this `sendUpdate` (note that I've removed `sendUpdate` in my codesandbox altogether).

It's also more error-prone because you don't really listen for the changes in the child, you just hope to read this state at a correct point in time. Changes in the child won't actually be automatically reflected in your component because your component is not actually subscribed to this child actor.

**Comment by Andarist** [comment]
It's also quite interesting that your original codesandbox gets magically fixed if we switch it to using `ReactDOM.createRoot` API (and not the classic `ReactDOM.render`):
https://codesandbox.io/s/xstate-4-33-2-sendupdate-error-forked-ipxh86?file=/src/index.js

It's because with `createRoot` updates are batched and even though there is no other update happening here, React doesn't know it so it just has to wait till the end of the stack to see if there is anything to be batched. So when it finally decides to flush the render (cause there won't be any other updates coming in) we have already updated the child's internal state and your code works. I don't truly recommend this is a solution but I've found this to be interesting 😅
